src/plugins/mmapstorage/mmapstorage.c

```diff
@@ -0,0 +1,1323 @@
+				}
+			}
+		}
+	}
+
+	// allocate space in DynArray to remember the addresses of meta keys
+	if (dynArray->size > 0)
+	{
+		dynArray->mappedKeyArray = elektraCalloc (dynArray->size * sizeof (Key *));
+	}
+
+	// duplicate all unique meta keys
+	Key * copy = 0;
+	cur = 0;
+	for (size_t i = 0; i < dynArray->size; ++i)
+	{
+		cur = dynArray->keyArray[i];
+		copy = keyDup (cur);
+
+		dynArray->mappedKeyArray[i] = copy;
+	}
+
+	// duplicate keyset from mmap and replace meta keys with deep copy
+	KeySet * dest = mmapKsDeepDup (toCopy);
+	cur = 0;
+	ksRewind (dest);
+	while ((cur = ksNext (dest)) != 0)
+	{
+		if (cur->meta)
+		{
+			KeySet * newMetaKs = ksNew (cur->meta->alloc, KS_END);
+
+			ksRewind (cur->meta);
+			for (size_t i = 0; i < cur->meta->size; ++i)
+			{
+				Key * newMetaKey =
+					dynArray->mappedKeyArray[elektraMmapDynArrayFind ((Key *) cur->meta->array[i], dynArray)];
+				ksAppendKey (newMetaKs, newMetaKey);
+			}
+			ksDel (cur->meta);
+			cur->meta = newMetaKs;
+		}
+	}
+
+	elektraMmapDynArrayDelete (dynArray);
+	return (KeySet *) dest;
+}
+
+/**
+ * @brief Unlink a mapped file.
+ *
+ * Before a mapped file can be altered, all mapped keysets have to be copied out
+ * from the mapped region. This approach is called unlinking a mapped file.
+ *
+ * To be able to unlink mapped keysets, the plugin data holds information with pointers
+ * of mapped keysets and the corresponding files. The Key names contain the filenames and
+ * mappings of a file are stored within meta-keys.
+ *
+ * Whether a keyset has been deleted has to be checked using keyset->mmapMetaData->flags
+ * (MMAP_FLAG_DELETED).
+ *
+ * @param parentKey holding the filename of the mapped file.
+ */
+static void unlinkFile (Key * parentKey)
+{
+	ELEKTRA_LOG_DEBUG ("unlink: need to unlink old mapped memory from file");
+
+	const Key * cur;
+	keyRewindMeta (parentKey);
+	ELEKTRA_LOG_DEBUG ("unlink: file: %s", keyString (parentKey));
+	while ((cur = keyNextMeta (parentKey)) != 0)
+	{
+		void * toUnlinkMmap = hexStringToAddress (keyName (cur));
+		ELEKTRA_LOG_DEBUG ("unlink: unlinking mmap str: %s", keyName (cur));
+		ELEKTRA_LOG_DEBUG ("unlink: unlinking mmap ptr: %p", toUnlinkMmap);
+
+		void * toUnlinkMmapMetaData = (char *) toUnlinkMmap + SIZEOF_MMAPHEADER;
+		ELEKTRA_LOG_DEBUG ("unlink: unlinking mmap meta-data ptr: %p", toUnlinkMmapMetaData);
+
+		void * toUnlinkKS = hexStringToAddress (keyString (cur));
+		ELEKTRA_LOG_DEBUG ("unlink: unlinking KeySet str: %s", keyString (cur));
+		ELEKTRA_LOG_DEBUG ("unlink: unlinking KeySet ptr: %p", toUnlinkKS);
+
+		if (!toUnlinkMmap || !toUnlinkKS)
+		{
+			ELEKTRA_LOG_DEBUG ("unlink: skipping unlinking, some pointers were invalid");
+			continue;
+		}
+
+		KeySet * copy = copyKeySet (toUnlinkKS, toUnlinkMmapMetaData);
+		if (copy)
+		{
+			KeySet * keySet = (KeySet *) toUnlinkKS;
+			ksClose (keySet);
+			keySet->array = copy->array;
+			keySet->size = copy->size;
+			keySet->alloc = copy->alloc;
+			keySet->cursor = 0;
```

Comment:
Shouldn't that be that this should be done ksClose? (It seems like currently it is not done and the docu of ksClose is not very enlightening, too :cry:)

Reply:
You're right. ksClose calls ksRewind which does exactly that. I removed my unnecessary manual rewinds.